From 891d8ff98e352f04fcd863a79e5d8ae00198db0f Mon Sep 17 00:00:00 2001
From: Mats Erik Andersson <gnu@gisladisker.se>
Date: Sat, 23 Jan 2016 22:47:41 +0100
Subject: [PATCH 19/42] ftp: Allow remote user in command.

Implement the extended host argument form `user@host'
for quick selection of the remote user name.
---
 ChangeLog          | 12 ++++++++++++
 NEWS               |  5 +++++
 doc/inetutils.texi |  9 +++++++++
 ftp/ftp.c          | 17 +++++++++++++++--
 4 files changed, 41 insertions(+), 2 deletions(-)

diff --git a/ftp/ftp.c b/ftp/ftp.c
index 11d4adc6..ff0b86cd 100644
--- a/ftp/ftp.c
+++ b/ftp/ftp.c
@@ -133,7 +133,11 @@ hookup (char *host, int port)
   int s, tos;
   socklen_t len;
   static char hostnamebuf[80];
-  char *rhost;
+  char *rhost, *p;
+
+  p = strrchr (host, '@');
+  if (p && p != host && isprint (p[1]))
+    host = ++p;
 
 #ifdef HAVE_IDN
   status = idna_to_ascii_lz (host, &rhost, 0);
@@ -293,7 +297,7 @@ int
 login (char *host)
 {
   char tmp[80];
-  char *user, *pass, *acct;
+  char *user, *pass, *acct, *p;
   int n, aflag = 0;
 
   user = pass = acct = 0;
@@ -302,6 +306,15 @@ login (char *host)
       code = -1;
       return (0);
     }
+
+  p = strrchr (host, '@');
+  if (user == NULL && p && p != host && isprint (p[1]))
+    {
+      *p = 0;
+      user = host;
+      host = ++p;
+    }
+
   while (user == NULL)
     {
       char *myname = getlogin ();
-- 
2.20.1.791.gb4d0f1c61a

